sycl-fusion/passes/kernel-fusion/SYCLKernelFusion.cpp

```diff
@@ -0,0 +1,795 @@
+}
+
+struct FunctionCall {
+  FunctionCallee Callee;
+  SmallVector<Value *> Args;
+};
+
+template <typename T> static void setBarrierMetadata(T *Ptr, LLVMContext &C) {
+  const auto FnAttrs =
+      AttributeSet::get(C, {Attribute::get(C, Attribute::AttrKind::Convergent),
+                            Attribute::get(C, Attribute::AttrKind::NoUnwind)});
+  Ptr->setAttributes(AttributeList::get(C, FnAttrs, {}, {}));
+  Ptr->setCallingConv(CallingConv::SPIR_FUNC);
+}
+
+static FunctionCall createBarrierCall(IRBuilderBase &Builder, Module &M,
+                                      int Flags) {
+  assert((Flags == 1 || Flags == 2 || Flags == 3) && "Invalid barrier flags");
+
+  constexpr StringLiteral N{"_Z22__spirv_ControlBarrierjjj"};
+
+  llvm::Function *F = M.getFunction(N);
+  if (!F) {
+    constexpr auto Linkage = GlobalValue::LinkageTypes::ExternalLinkage;
+
+    auto *Ty = FunctionType::get(
+        Builder.getVoidTy(),
+        {Builder.getInt32Ty(), Builder.getInt32Ty(), Builder.getInt32Ty()},
+        false /* isVarArg*/);
+
+    F = Function::Create(Ty, Linkage, N, M);
+
+    setBarrierMetadata(F, Builder.getContext());
+  }
+
+  // See
+  // https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Memory_Semantics_-id-
+  return {F,
+          {Builder.getInt32(/*Exec Scope : Workgroup = */ 2),
+           Builder.getInt32(/*Exec Scope : Workgroup = */ 2),
+           Builder.getInt32(0x10 | (Flags % 2 == 1 ? 0x100 : 0x0) |
+                            ((Flags >> 1 == 1 ? 0x200 : 0x0)))}};
+}
+
+void SYCLKernelFusion::fuseKernel(
+    llvm::Module &M, llvm::Function &StubFunction,
+    jit_compiler::SYCLModuleInfo *ModInfo,
+    SmallPtrSetImpl<llvm::Function *> &ToCleanUp) const {
+  // Retrieve the metadata from the stub function.
+  // The first operand of the tuple is the name that the newly created,
+  // fused function should carry.
+  // The second operand is the list (MDTuple) of function names that should be
+  // fused into this fused kernel.
+  // Fuction names may appear multiple times, resulting in
+  // the corresponding function to be included multiple times. The functions
+  // are included in the same order as given by the list.
+  llvm::MDNode *MD = StubFunction.getMetadata(MetadataKind);
+  assert(MD && MD->getNumOperands() == 2 &&
+         "Metadata tuple should have two operands");
+
+  // Name of the fused kernel.
+  auto *KernelName = dyn_cast<MDString>(MD->getOperand(0).get());
+  assert(KernelName &&
+         "First metadata operand should be the fused kernel name");
+
+  // List of kernels to fuse into this function.
+  auto *KernelList = dyn_cast<MDNode>(MD->getOperand(1).get());
+  assert(KernelList &&
+         "Second metadata operand should be the list of kernels to fuse");
+  SmallVector<StringRef> FusedKernels;
+  for (const llvm::MDOperand &MDOp : KernelList->operands()) {
+    auto *FK = dyn_cast<MDString>(MDOp.get());
+    assert(FK && "Kernel should be given by its name as MDString");
```

Comment:
No need to use `dyn_cast` if you don't check the result. `cast` already contains an `assert` statement for you